# Assignment 2

In this assignment you will implement two ways of performing automatic differentiation: forward accumulation and reverse accumulation.

## Part A: Automatic Differentiation with Forward Accumulation

In forward accumulation, in one pass the algorithm evaluates the expression and calculates its derivative with respect to one input variable. However, if there are multiple input variables, then the algorithm has to be called once for each input variable.

#### Provided sample classes:
We provide some sample code and examples in the following for reference. Specifically, we provide the expression base class and common operator with its derivative functions.

In [84]:
import math
import numpy as np

class Expression:
    def __init__(self, children=None, op=''):
        self.children = children or []
        self.op = op

    def __add__(self, other):
        return AddExpression(self, other)

    def __mul__(self, other):
        return MulExpression(self, other)

    def __truediv__(self, other):
        return TruedivExpression(self, other)
    
    def __pow__(self, other):
        return PowExpression(self, other)
    
    def evaluate_derive(self, variable):
        raise Exception('Method evaluate_derive not implemented: ' + self.op)

    def __repr__(self):
        return f'({self.op}, ' + str(self.children) + ')'

class Var(Expression):
    def __init__(self, value, op='Var'):
        self.value = value
        self.children = []
        self.op = op
        
    def evaluate_derive(self, variable):
        if variable == self:
            return [self.value, 1]
        else:
            return [self.value, 0]

    def __repr__(self):
        return f'({self.op}, ' + str(self.value) + ')'

class sin(Expression):
    def __init__(self, expr):
        super().__init__([expr], 'sin')
    def evaluate_derive(self, variable):
        value, grad = self.children[0].evaluate_derive(variable)
        return [math.sin(value), math.cos(value) * grad]

class sigmoid(Expression):
    def __init__(self, expr):
        super().__init__([expr], 'sigmoid')
    def evaluate_derive(self, variable):
        
        ######## Insert your code below #########
        
        value, grad = self.children[0].evaluate_derive(variable)
        return [1/(1+math.exp(-value)), 1/(1+math.exp(-value))*(1 - (1/(1+math.exp(-value))))*grad]
        
        #raise NotImplementedError

        
        
        
        ##########################################
        
        

class AddExpression(Expression):
    def __init__(self, left, right):
        super().__init__([left, right], 'add')
        
    def evaluate_derive(self, variable):
        value_left, grad_left = self.children[0].evaluate_derive(variable)
        value_right, grad_right = self.children[1].evaluate_derive(variable)
        return [value_left + value_right, grad_left + grad_right]

class MulExpression(Expression):
    def __init__(self, left, right):
        super().__init__([left, right], 'mul')
    
    def evaluate_derive(self, variable):
        value_left, grad_left = self.children[0].evaluate_derive(variable)
        value_right, grad_right = self.children[1].evaluate_derive(variable)
        return [
            value_left * value_right,
            value_right * grad_left + value_left * grad_right]

class cos(Expression):
    def __init__(self, expr):
        super().__init__([expr], 'cos')
    def evaluate_derive(self, variable):
        ######## Insert your code below #########
        
        value, grad = self.children[0].evaluate_derive(variable)
        return [math.cos(value), -math.sin(value)*grad]
        #raise NotImplementedError
       

        
        ##########################################

    
class TruedivExpression(Expression):
    def __init__(self, left, right):
        super().__init__([left, right], 'div')
    def evaluate_derive(self, variable):
        
        ######## Insert your code below #########
        
        value_left, grad_left = self.children[0].evaluate_derive(variable)
        value_right, grad_right = self.children[1].evaluate_derive(variable)
        return [(value_left/value_right), ((grad_left*value_right - value_left*grad_right)/(value_right**2))]
        #raise NotImplementedError

        
        
        #########################################

class PowExpression(Expression):
    def __init__(self, left, right):
        super().__init__([left, right], 'pow')
        
    def evaluate_derive(self, variable):

        ######## Insert your code below #########
        
        value_left, grad_left = self.children[0].evaluate_derive(variable)
        value_right, grad_right = self.children[1].evaluate_derive(variable)
        return [(value_left**value_right), ((value_left**value_right)*((grad_right * np.log(value_left)) + (value_right*(grad_left/value_left))))]
        #raise NotImplementedError

        
        ##########################################      

class exp(Expression):
    def __init__(self, expr):
        super().__init__([expr], 'exp')
    def evaluate_derive(self, variable):
        value, grad = self.children[0].evaluate_derive(variable)
        return [np.exp(value), math.exp(value) * grad]

class log(Expression):
    def __init__(self, expr):
        super().__init__([expr], 'log')
    def evaluate_derive(self, variable):
        
        ######## Insert your code below #########
        
        value, grad = self.children[0].evaluate_derive(variable)
        return [np.log(value), (grad/value)]
        #raise NotImplementedError
         
        ##########################################



#### Example 1. Define $x=2$ and $y=e^x$. Find $\frac{\partial{y}}{\partial{x}}|_{x=2}$.

In [85]:
x = Var(2, op='x')
y = exp(x)
print(y, y.evaluate_derive(x))

(exp, [(x, 2)]) [7.38905609893065, 7.38905609893065]


Output [7.38905609893065, 7.38905609893065] indicates that $y=7.8390$ and $\frac{\partial{y}}{\partial{x}}|_{x=2}=7.8390$.

#### <span style="color:red">Question 1 (3pt): </span> Define $x=3$ and $z=\sin(x)$. Find $\frac{\partial{z}}{\partial{x}}|_{x=3}$.

In [88]:
######## Insert your code below #########
        
x = Var(3, op = 'x')
z = sin(x)
print(z, z.evaluate_derive(x))

        
        
#########################################

(sin, [(x, 3)]) [0.1411200080598672, -0.9899924966004454]


#### <span style="color:red">Question 2 (10pt): </span> Implement $\log(x)$ and $\cos(x)$ function in the sample code provided at the beginning of Part A. Please do not modify any given functions and only insert your code where indicated.

#### Example 2. Define $a=3$, $b=4$, $x=\log(a)$, $y=\cos(b)$ and $z=x + y$. Find $\frac{\partial{z}}{\partial{a}}|_{a=3, b=4}$ and $\frac{\partial{z}}{\partial{b}}|_{a=3,b=4}$. If your implmentation of $log(x)$ and $cos(x)$ are correct, you should get $\frac{\partial{z}}{\partial{a}}|_{a=3}=0.3333$ and $\frac{\partial{z}}{\partial{b}}|_{b=4}=0.7568$.

In [89]:
a = Var(3, op='a')
b = Var(4, op='b')
x = log(a)
y = cos(b)
z = x + y
print(z, z.evaluate_derive(a))
print(z, z.evaluate_derive(b))

(add, [(log, [(a, 3)]), (cos, [(b, 4)])]) [0.44496866780449784, 0.3333333333333333]
(add, [(log, [(a, 3)]), (cos, [(b, 4)])]) [0.44496866780449784, 0.7568024953079282]


#### <span style="color:red">Question 3 (3pt): </span> Define $x=4$ and $y=\log(x+4)$. Find $\frac{\partial{y}}{\partial{x}}|_{x=4}$.

In [90]:
######## Insert your code below #########
               
x = Var(4, op = 'x')
const = Var(4, op = 'cons')
y = log(x + const)
print(y, y.evaluate_derive(x))
        
        
#########################################

(log, [(add, [(x, 4), (cons, 4)])]) [2.0794415416798357, 0.125]


#### <span style="color:red">Question 4 (10pt): </span> Implement the power function $x^y$ and the sigmoid function in the sample code provided at the beginning of Part A. The sigmoid function is defined the following way: $\sigma(x)=\dfrac{1}{1+e^{-x}}$. Please do not modify any given functions and only insert your code where indicated.

#### Example 3. Define $x=0.1$, $y=0.2$, $z=x^y$, and $w=\sigma(z)$. Find $\frac{\partial{w}}{\partial{x}}|_{x=0.1, y=0.2}$. If your implementation of $\sigma(x)$ is correct, you should get $\frac{\partial{w}}{\partial{x}}|_{x=0.1, y=0.2}=0.2860$ .

In [91]:
x = Var(0.1, op='x')
y = Var(0.2, op='y')
z = x ** y
w = sigmoid(z)
print(w, w.evaluate_derive(x))

(sigmoid, [(pow, [(x, 0.1), (y, 0.2)])]) [0.65270650544543, 0.28605173430348574]


#### <span style="color:red">Question 5 (10pt): </span> Implement the class TruedivExpression in the sample code provided at the beginning of Part A. You can use the provided class AddExpression, MulExpression as examples. Please do not modify any given functions and only insert your code where indicated.

#### Example 4. Define $x=\pi$, and $y=\pi^2 \cdot \frac{\sin(x)}{x^2}$. Find $\frac{\partial{y}}{\partial{x}}|_{x=\pi}$. If your implementation of class TruedivExpression is correct, you should get $\frac{\partial{y}}{\partial{x}}|_{x=\pi}=-1$ .

In [92]:
x = Var(math.pi, op='x')
c1 = Var(2, op='c1')
c2 = Var(math.pi ** 2, op='c2')
y = c2 * sin(x) / (x ** c1)
print(y, y.evaluate_derive(x))

(div, [(mul, [(c2, 9.869604401089358), (sin, [(x, 3.141592653589793)])]), (pow, [(x, 3.141592653589793), (c1, 2)])]) [1.2246467991473532e-16, -1.0000000000000002]


#### <span style="color:red">Question 6 (6pt): </span> Define $x=0.5$, $y=1.5$ and $z= x \cdot y + \sin(x)\cdot \cos(x) \cdot \log(x \cdot y) + \sigma(y) \cdot \log_2(x+y)$. Find $\frac{\partial{z}}{\partial{y}}|_{x=0.5, y=1.5}$. 

In [93]:
######## Insert your code below #########

x = Var(0.5, op = 'x')
y = Var(1.5, op = 'y')
const= Var(2, op = 'const')
z = x*y + sin(x)*cos(x)*log(x*y)+sigmoid(y)*(log(x+y)/log(const))
print(z, z.evaluate_derive(y))

        
        
#########################################

(add, [(add, [(mul, [(x, 0.5), (y, 1.5)]), (mul, [(mul, [(sin, [(x, 0.5)]), (cos, [(x, 0.5)])]), (log, [(mul, [(x, 0.5), (y, 1.5)])])])]), (mul, [(sigmoid, [(y, 1.5)]), (div, [(log, [(add, [(x, 0.5), (y, 1.5)])]), (log, [(const, 2)])])])]) [1.4465364177848552, 1.5193921015206122]


#### <span style="color:red">Question 7 (8pt): </span> Define $x=1$, $y=2$, $z=3$ and $w=\sin(\frac{1}{x})\cdot \log_{10}(y^3)+\cos(\frac{\sqrt{z}}{\exp(y)})$. Find $\frac{\partial{w}}{\partial{x}}|_{x=1,y=2,z=3}$, $\frac{\partial{w}}{\partial{y}}|_{x=1,y=2,z=3}$ and $\frac{\partial{w}}{\partial{z}}|_{x=1,y=2,z=3}$.

In [94]:
######## Insert your code below #########

x = Var(1, op = 'x')
y = Var(2, op = 'y')
z = Var(3, op = 'z')
const1 = Var(1, op = 'const1')
const2 = Var(3, op = 'const2')
const3 = Var(10, op = 'const3')
const4 = Var(0.5, op = 'const4')
w = sin(const1/x) * (log(y**const2)/log(const3)) + cos((z**const4)/exp(y))
print(w, w.evaluate_derive(x), w.evaluate_derive(y), w.evaluate_derive(z))
        
        
#########################################

(add, [(mul, [(sin, [(div, [(const1, 1), (x, 1)])]), (div, [(log, [(pow, [(y, 2), (const2, 3)])]), (log, [(const3, 10)])])]), (cos, [(div, [(pow, [(z, 3), (const4, 0.5)]), (exp, [(y, 2)])])])]) [1.7325761306945973, -0.48794160237817535] [1.7325761306945973, 0.6026144114405535] [1.7325761306945973, -0.009074183894151131]


#### Appendix: How to verify your implmentation using pytorch?
PyTorch has a well-known built-in automatic differentiation system. Below, we provide a simple script demonstrating how to use it in a basic 1-dimensional case. Please note that this is a fundamental example; PyTorch’s automatic differentiation is much more powerful and can handle tensors of any dimension.

Ensure that you run the following code only after completing all the required implementations; otherwise, you may encounter errors.

#### Define $x=5$, $y=3$, $a=x^y$, $b=x \cdot \sigma(a)$ and $z = (\frac{a}{b})^2$. Find $\frac{\partial{z}}{\partial{x}}|_{x=5,y=3}$ and $\frac{\partial{z}}{\partial{y}}|_{x=5,y=3}$.

In [95]:
## PyTorch implementation 

import torch
x = torch.tensor(5.0, requires_grad=True)
y = torch.tensor(3.0, requires_grad=True)
a = x.pow(y)
b = x * torch.sigmoid(a)
z = (a / b).pow(2)
z.backward()
print(z, x.grad, y.grad)

## Our forward implementation

x = Var(5.0, op='x')
y = Var(3.0, op='y')
a = x ** y
b = x * sigmoid(a)
z = (a / b) ** (Var(2.0, op='c'))
print(z.evaluate_derive(x))
print(z.evaluate_derive(y))

## Same results

tensor(625., grad_fn=<PowBackward0>) tensor(500.) tensor(2011.7975)
[625.0, 500.0000000000001]
[625.0, 2011.7973905426254]


## Part B: Automatic Differentiation with Reverse Accumulation

In reverse accumulation, the algorithm performs two passes. In the first pass it evaluates the expression, then in the second pass it computes all the partial derivatives with respect to all the input variables. Reverse accumulation is faster if there more input variables than output variables. However, compared to forward accumulation reverse accumulation does require more memory because it has to store the results of the first pass.

#### Provided sample classes:
We provide some sample code and examples in the following for reference. Specifically, we provide the expression base class and common operator with its derivative functions.

In [96]:
import math
import numpy as np

class Var:
    def __init__(self, value, children=None, op='Var'):
        self.value = value
        self.grad = 0
        self.children = children or []
        self.op = op

    def __add__(self, other):
        return Var(
            self.value + other.value,
            [(1, self), (1, other)],
            'add')

    def __mul__(self, other):
        return Var(
            self.value * other.value,
            [(other.value, self), (self.value, other)],
            'mul')

    def __truediv__(self, other):
        ######## Insert your code below #########
        
        return Var(self.value/other.value, 
                   [(1/other.value, self), (-self.value/other.value**2, other)], 
                   'div')
        #raise NotImplementedError
      
        
        
        ########################################################
        


    def __pow__(self, other):
        
        ######## Remove pass and insert your code below #########
        
        return Var(self.value**other.value, 
                   [(other.value*self.value**(other.value-1), self), ((self.value**other.value)*np.log(self.value), other)],
                   'pow')
        #raise NotImplementedError
        
        
        #########################################################

        
        
    def backward(self, grad=1):
        # print('backward:', self.op, grad)
        self.grad += grad
        for coef, child in self.children:
            child.backward(grad * coef)

    def __repr__(self):
        return f'({self.op}, ' + str(self.value) + ')'

class sin(Var):
    def __init__(self, expr):
        ######## Remove pass and insert your code below #########
        
        super().__init__(
            np.sin(expr.value),
            [(np.cos(expr.value), expr)],
            type(self).__name__)
        #raise NotImplementedError
        
        
        
        #########################################################
        
class exp(Var):
    def __init__(self, expr):
        super().__init__(
            np.exp(expr.value),
            [(np.exp(expr.value), expr)],
            type(self).__name__)
        
class log(Var):
    def __init__(self, expr):
        ######## Remove pass and insert your code below #########
        
        super().__init__(
            np.log(expr.value),
            [(1/(expr.value), expr)],
            type(self).__name__)
       #raise NotImplementedError

        
        
        #########################################################
        
class cos(Var):
    def __init__(self, expr):
        super().__init__(
            np.cos(expr.value),
            [(-np.sin(expr.value), expr)],
            type(self).__name__)
        
class tanh(Var):
    def __init__(self, expr):
        ######## Remove pass and insert your code below #########
        
        super().__init__(
            np.tanh(expr.value),
            [(1 - np.tanh(expr.value)**2, expr)],
            type(self).__name__)
        #raise NotImplementedError

        
        #########################################################
        

#### Example 1. Define $x=3$ and $z=\cos(x)$. Find $\frac{\partial{z}}{\partial{x}}|_{x=3}$.

In [97]:
x = Var(3, op='x')
z = cos(x)
z.backward()
print(z, x.grad)

(cos, -0.9899924966004454) -0.1411200080598672


Output (cos, -0.9899924966004454) -0.1411200080598672 indicates that $z=1.0986$ and $\frac{\partial{z}}{\partial{x}}|_{x=3}=-0.1411$.

#### <span style="color:red">Question 1 (7pt): </span> Define $x=0.5$ and $y=\exp(x) + \cos(x)$. Find $\frac{\partial{y}}{\partial{x}}|_{x=0.5}$.

In [98]:
######## Insert your code below #########

x = Var(0.5, op = 'x')
y = exp(x) + cos(x)
y.backward()
print(y, x.grad)

        
        
#########################################

(add, 2.526303832590501) 1.1692957320959252


#### <span style="color:red">Question 2 (10pt): </span> Implement the $\log(x)$ and $\sin(x)$ function in the sample code provided at the beginning of Part B. Please do not modify any given functions and only insert your code where indicated. 

#### Example 2. Define $x=3$, $y=2$ and $z=\log(\cos(\exp(x))\cdot \sin(y))$. Find $\frac{\partial{z}}{\partial{x}}|_{x=3,y=2}$ and $\frac{\partial{z}}{\partial{y}}|_{x=3,y=2}$. If your implementation of $\log(x)$ and $\sin(x)$ are correct, you should get $\frac{\partial{z}}{\partial{x}}|_{x=3,y=2}=-57.7313$ and $\frac{\partial{z}}{\partial{y}}|_{x=3,y=2}=-0.4576$.

In [99]:
x = Var(3, op = 'x')
y = Var(2, op = 'y')
z = log(cos(exp(x)) * sin(y))
z.backward()
print(z, x.grad, y.grad)


(log, -1.208013069940718) -57.73131496788059 -0.45765755436028577


#### <span style="color:red">Question 3 (10pt): </span> Implement the power function $x^y$ and the hyperbolic tangent function in the sample code provided at the beginning of Part B. The hyperbolic tangent function is defined the following way: $\tanh(x)=\dfrac{\exp(x)-\exp(-x)}{\exp(x)+\exp(-x)}$. 

#### Example 3. Define $x=0.5$ and $z=x^x \cdot \tanh(x)$. Find $\frac{\partial{z}}{\partial{x}}|_{x=0.5}$. If your implementation of $\sigma(x)$ is correct, you should get $\frac{\partial{z}}{\partial{x}}|_{x=0.5}=0.6563$ .

In [100]:
x = Var(0.5, op='x')
z = (x ** x) * tanh(x)
z.backward()
print(z, x.grad)

(mul, 0.3267661756012031) 0.6563716473098675


#### <span style="color:red"> Question 4 (10pt):</span> Implement the function \_\_truediv\_\_ in the sample code provided at the beginning of Part B. You can use the provided class AddExpression, MulExpression as examples. Please do not modify any given functions and only insert your code where indicated.

#### Example 4. Define $x=0$ and $z=\tan(x)=\frac{\sin(x)}{\cos(x)}$. Find $\frac{\partial{z}}{\partial{x}}|_{x=0}$. If your implementation of $\sigma(x)$ is correct, you should get $\frac{\partial{z}}{\partial{x}}|_{x=0}=1$ .

In [101]:
x = Var(0, op='x')
z = sin(x) / cos(x)
z.backward()
print(z, x.grad)

(div, 0.0) 1.0


#### <span style="color:red">Question 5 (8pt): </span> Define $x=1$, $y=2$, $z=3$ and $w=\sin(\frac{1}{x})\cdot \log_{10}(y^3)+\cos(\frac{\sqrt{z}}{\exp(y)})$. Find $\frac{\partial{w}}{\partial{x}}|_{x=1,y=2,z=3}$, $\frac{\partial{w}}{\partial{y}}|_{x=1,y=2,z=3}$ and $\frac{\partial{w}}{\partial{z}}|_{x=1,y=2,z=3}$. 

In [102]:
######## Insert your code below #########

x = Var(1, op = 'x')
y = Var(2, op = 'y')
z = Var(3, op = 'z')
const1 = Var(1, op = 'const1')
const2 = Var(3, op = 'const2')
const3 = Var(10, op = 'const3')
const4 = Var(0.5, op = 'const4')
w = sin(const1/x) * (log(y**const2)/log(const3)) + cos((z**const4)/exp(y))
w.backward()
print(w, x.grad, y.grad, z.grad)

        
        
#########################################

(add, 1.7325761306945973) -0.48794160237817535 0.6026144114405535 -0.009074183894151133


#### <span style="color:red">Question 6 (5pt): </span> Compare with Question 7 in Part A, briefly discuss what are the differences between them. (Hint: which one is faster or more efficient?)

**Your discussion starts here:**
Backward pass goes from output to input while forward pass goes from input to output. For forward pass, the model parameters are used to compute both the intermediate input and output variables while during backward pass, the intermediate variables are used to find the gradient. Therefore, in cases where there's a greater number of inputs than outputs such as Question 5 above, the backward pass would be more efficeint while for cases where there's more outputs than inputs, forward pass would be more efficient. In cases where there's an equal amount of inputs and outputs, they may be equally efficient.

#### Appendix: How to verify your implmentation using pytorch?
PyTorch has a well-known built-in automatic differentiation system. Below, we provide a simple script demonstrating how to use it in a basic 1-dimensional case. Please note that this is a fundamental example; PyTorch’s automatic differentiation is much more powerful and can handle tensors of any dimension.

Ensure that you run the following code only after completing all the required implementations; otherwise, you may encounter errors.

#### Define $x=5$, $y=3$, $a=x^y$, $b=x \cdot \tanh(a)$ and $z = (\frac{a}{b})^2$. Find $\frac{\partial{z}}{\partial{x}}|_{x=5,y=3}$ and $\frac{\partial{z}}{\partial{y}}|_{x=5,y=3}$.

In [103]:
## PyTorch implementation 

import torch
xt = torch.tensor(5.0, requires_grad=True)
yt = torch.tensor(3.0, requires_grad=True)
at = xt.pow(yt)
bt = xt * torch.tanh(at)
zt = (at / bt).pow(2)
zt.backward()
print(zt, xt.grad, yt.grad)

## Our backward implementation

x = Var(5.0, op='x')
y = Var(3.0, op='y')
a = x ** y
b = x * tanh(a)
z = (a / b) ** (Var(2.0, op='c'))
z.backward()
print(z, x.grad, y.grad)

## Same results

tensor(625., grad_fn=<PowBackward0>) tensor(500.) tensor(2011.7975)
(pow, 625.0) 500.0 2011.7973905426254
